### PR TITLE
Cuda block size and function overload

### LINKED ISSUE #2066: cudaFuncSetCacheConfig is setting CachePreferShared too often
Is there a way that a Kokkos client can call cudaFuncSetCacheConfig? Right now, it appears that CudaParallelLaunch always calls it on new GPUs, and it also appears that it always passes cudaFuncCachePreferShared for TeamPolicy, even if the user does not request shared memory. I have found in some kernels that if I mod that call in the Kokkos source code to prefer L1, the kernels run substantially faster.

If there is not a way, I suggest TeamPolicy take something like Kokkos::CachePreferShared or Kokkos::CachePreferL1 as an option.

### PR DESCRIPTION
**Pull request by crtrott**
This fixes #2066 and #2127

### CODE REVIEW

**Comment by crtrott** [comment]
@dalg can you take a look why the Jenkins thing failed? 

```
Using context: continuous-integration/jenkin-ci/pr
ERROR: /var/jenkins_home/jobs/Kokkos/workspace@script/.jenkins not found
Finished: FAILURE
```

**Comment by crtrott** [comment]
Spot Check is ok
```
Running on machine: apollo
Repository Status:  7d6465df2d9f22571f4991c5aefe5c9767f0e7f7 Fix unsigned/signed warning in Clang/CUDA


sems-cmake/3.5.2(3):ERROR:150: Module 'sems-cmake/3.5.2' conflicts with the currently loaded module(s) 'sems-cmake/3.12.2'
sems-cmake/3.5.2(3):ERROR:102: Tcl command execution failed: conflict sems-cmake

Going to test compilers:  gcc/4.8.4 gcc/5.3.0 intel/16.0.1 clang/3.9.0 clang/6.0 cuda/9.1 hpx/1.2.1
Testing compiler gcc/4.8.4
  Starting job gcc-4.8.4-OpenMP-release
  PASSED gcc-4.8.4-OpenMP-release
Testing compiler gcc/5.3.0
  Starting job gcc-4.8.4-Pthread-release
  PASSED gcc-4.8.4-Pthread-release
Testing compiler intel/16.0.1
  Starting job gcc-5.3.0-Serial-release
  PASSED gcc-5.3.0-Serial-release
Testing compiler clang/3.9.0
  Starting job intel-16.0.1-OpenMP-release
  PASSED intel-16.0.1-OpenMP-release
Testing compiler clang/6.0
  Starting job clang-3.9.0-Pthread_Serial-release
  PASSED clang-3.9.0-Pthread_Serial-release
  Starting job clang-6.0-Cuda_Pthread-release
  PASSED clang-6.0-Cuda_Pthread-release
Testing compiler cuda/9.1
  Starting job clang-6.0-OpenMP-release
  PASSED clang-6.0-OpenMP-release
Testing compiler hpx/1.2.1
  Starting job cuda-9.1-Cuda_OpenMP-release
  PASSED cuda-9.1-Cuda_OpenMP-release
  Starting job hpx-1.2.1-HPX-release
  PASSED hpx-1.2.1-HPX-release
#######################################################
PASSED TESTS
#######################################################
clang-3.9.0-Pthread_Serial-release build_time=135 run_time=476
clang-6.0-Cuda_Pthread-release build_time=241 run_time=333
clang-6.0-OpenMP-release build_time=115 run_time=94
cuda-9.1-Cuda_OpenMP-release build_time=301 run_time=196
gcc-4.8.4-OpenMP-release build_time=106 run_time=128
gcc-4.8.4-Pthread-release build_time=97 run_time=389
gcc-5.3.0-Serial-release build_time=114 run_time=453
hpx-1.2.1-HPX-release build_time=767 run_time=355
intel-16.0.1-OpenMP-release build_time=346 run_time=152
```

**Comment by Dalg** [comment]
What is this?

**Comment by crtrott** [comment]
@dalg sorry I mentioned the wrong person ...

@dalg24 Can you take a look at this?

File: core/src/Cuda/Kokkos_Cuda_Parallel.hpp [hunk 1]
```diff
@@ -749,25 +769,35 @@ class ParallelFor< FunctorType
...
+
       // Functor's reduce memory, team scan memory, and team shared memory depend upon team size.
-      m_scratch_ptr[1] = cuda_resize_scratch_space(static_cast<ptrdiff_t>(m_scratch_size[1])*static_cast<ptrdiff_t>(Cuda::concurrency()/(m_team_size*m_vector_size)));
+      m_scratch_ptr[0] = NULL;
+      m_scratch_ptr[1] = m_team_size<=0?NULL:cuda_resize_scratch_space(static_cast<ptrdiff_t>(m_scratch_size[1])*static_cast<ptrdiff_t>(Cuda::concurrency()/(m_team_size*m_vector_size)));
 
       const int shmem_size_total = m_shmem_begin + m_shmem_size ;
-      if ( CudaTraits::SharedMemoryCapacity < shmem_size_total ) {
+      if ( m_policy.space().impl_internal_space_instance()->m_maxShmemPerBlock < shmem_size_total ) {
+        printf("%i %i\n",m_policy.space().impl_internal_space_instance()->m_maxShmemPerBlock,shmem_size_total);
```
Comment (ndellingwood) [inline core/src/Cuda/Kokkos_Cuda_Parallel.hpp]: @crtrott is this intended to stay as extra info with the throw? If so can some context to the output values be added?

File: core/src/Cuda/Kokkos_Cuda_Parallel.hpp [hunk 2]
```diff
@@ -1512,35 +1542,30 @@ class ParallelReduce< FunctorType
...
 
     m_team_begin = UseShflReduction?0:cuda_single_inter_block_reduce_scan_shmem<false,FunctorType,WorkTag>( arg_functor , m_team_size );
     m_shmem_begin = sizeof(double) * ( m_team_size + 2 );
-    m_shmem_size = arg_policy.scratch_size(0,m_team_size) + FunctorTeamShmemSize< FunctorType >::value( arg_functor , m_team_size );
-    m_scratch_ptr[1] = cuda_resize_scratch_space(static_cast<std::int64_t>(m_scratch_size[1])*(static_cast<std::int64_t>(Cuda::concurrency()/(m_team_size*m_vector_size))));
+    m_shmem_size = m_policy.scratch_size(0,m_team_size) + FunctorTeamShmemSize< FunctorType >::value( arg_functor , m_team_size );
     m_scratch_size[0] = m_shmem_size;
-    m_scratch_size[1] = arg_policy.scratch_size(1,m_team_size);
+    m_scratch_size[1] = m_policy.scratch_size(1,m_team_size);
+    m_scratch_ptr[1] = m_team_size<=0?NULL:cuda_resize_scratch_space(static_cast<std::int64_t>(m_scratch_size[1])*(static_cast<std::int64_t>(Cuda::concurrency()/(m_team_size*m_vector_size))));
```
Comment (ndellingwood) [inline core/src/Cuda/Kokkos_Cuda_Parallel.hpp]: @crtrott if any more changes are made could `NULL` be replaced by `nullptr`? Otherwise we can update later

Comment (crtrott) [reply core/src/Cuda/Kokkos_Cuda_Parallel.hpp]: we leave it for now

File: core/src/impl/Kokkos_ViewMapping.hpp [hunk 3]
```diff
@@ -2956,7 +2956,8 @@ class ViewMapping< DstTraits , SrcTraits ,
     };
 
 public:
-
+  enum { is_assignable_data_type = is_assignable_value_type &&
```
Comment (ndellingwood) [inline core/src/impl/Kokkos_ViewMapping.hpp]: @crtrott I'm assuming this change will break Sacado, I'll update on the kokkos-promotion branch.

**Pull request merged**